pkg/controller/managedresources/merger.go

```diff
@@ -69,14 +93,63 @@ func mergeDeployment(scheme *runtime.Scheme, oldObj, newObj runtime.Object) erro
 	}
 
 	// Do not overwrite a Deployment's '.spec.replicas' if the new Deployment's '.spec.replicas'
-	// field is unset.
-	if newDeployment.Spec.Replicas == nil {
+	// field is unset or the Deployment is scaled by either an HPA or HVPA.
+	if newDeployment.Spec.Replicas == nil || preserveReplicas {
 		newDeployment.Spec.Replicas = oldDeployment.Spec.Replicas
 	}
 
+	mergePodTemplate(&oldDeployment.Spec.Template, &newDeployment.Spec.Template, preserveResources)
+
 	return scheme.Convert(newDeployment, newObj, nil)
 }
 
+func mergePodTemplate(oldPod, newPod *corev1.PodTemplateSpec, preserveResources bool) {
+	if !preserveResources {
+		return
+	}
+
+	// Do not overwrite a PodTemplate's resource requests / limits if it is scaled by an HVPA
+	for _, newContainer := range newPod.Spec.Containers {
+		newContainerName := newContainer.Name
+
+		for _, oldContainer := range oldPod.Spec.Containers {
+			oldContainerName := oldContainer.Name
+
+			if newContainerName == oldContainerName {
+				mergeContainer(&oldContainer, &newContainer, preserveResources)
+			}
+		}
+	}
+}
+
+func mergeContainer(oldContainer, newContainer *corev1.Container, preserveResources bool) {
+	if !preserveResources {
```

Comment:
It could/should be removed now, or?

Reply:
Hmm, I don't think so. In my opinion, this belongs to the `mergeContainer` func. The early exit I added to `mergePodTemplate` is just an optimization, because `mergePodTemplate` doesn't do anything else _yet_, and might be removed in the future, when we add additional logic there. But the `mergeContainer` func is still the one deciding on how to react on the different values of `preserveResources`. Besides that, the unit tests for example also test the `mergeContainer` func with `preserverResources` set to `false` and `true` (which is probably the correct way to test such a function, I guess).

Reply:
OK, thanks.